# Table of Contents

## Importing Libraries

## Importing Dataframes

## Reading Tasks
### - Grouping Data with pandas
### - Aggregating Data with agg()
### - Aggregating Data with transform()
### - Deriving Columns with loc()

## Task Questions
### - Step 1: Create New Notebook
### - Step 2: Aggregated Mean of Entire Dataframe
### - Step 3: Analyze Result and Compare with Subset Results
### - Step 4: Create Loyalty Flag for Existing Customers
### - Step 5: Difference of Spending Habits in Loyalty Flag
### - Step 6: Different Types of Spenders
### - Step 7: Frequent vs. Non-frequent Customers

## Exporting to Pickle File 

----

# Importing Libraries

In [1]:
import pandas as pd
import numpy as ny
import os

# Importing Dataframes

In [2]:
# Folder path
path= r'C:\Users\princess\Documents\09-2023 Instacart Basket Analysis'

In [3]:
# Importing 'orders_products_merged_update.pkl'

ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

# Reading Tasks

In [4]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                    NaN         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders

In [5]:
ords_prods_merge.shape

(32404859, 18)

In [6]:
# Creating a subset

df = ords_prods_merge[:1000000]

In [7]:
df.shape

(1000000, 18)

In [8]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                    NaN         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   
5                   19.0         196                  1          1   both   
6                   20.0         196                  1          1   both   
7                   14.0         196                  2          1   both   
8                    0.0         196                  4          1   both   
9                   30.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders  
5  Regularly busy   Regularly busy        Average orders  
6  Regularly busy     Busiest days           Most orders  
7  Regularly busy     Busiest days           Most orders  
8  Regularly busy     Busiest days           Most orders  
9      Least busy  Least busy days        Average orders

# Grouping Data with pandas

In [9]:
df.groupby('product_name')

In [10]:
# To recap, you should always use the groupby() function as part of a series of steps, namely, the following:

# 1. Split the data into groups based on some criteria.
# 2. Apply a function to each group separately.
# 3. Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.

# So far, you’ve only completed the first step—splitting the data into groups (with the group being the “product_name” column).
# Now, let’s take a look into the second step, which will involve some aggregation!

# Aggregating Data with agg()

In [11]:
#  If you were to calculate the mean of the “order_number” column grouped by the “department_id” column, you could quickly compare the average number of orders per user across each Instacart department. 

# This will involve two of the steps in the three-step process introduced above:

# 1. Split the data into groups based on “department_id.”
# 2. Apply the agg() function to each group to obtain the mean values for the “order_number” column.

### Performing a Single Aggregation

In [12]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [13]:
# Different way of finding the mean of order_number

df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

### Performing Multiple Aggregations

In [14]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Aggregating Data with transform()

In [15]:
# Now, let’s map this task onto the three-step process introduced earlier:

# 1. Split the data into groups based on the “user_id” column.
# 2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
# 3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.

In [16]:
# Calculating max number of orders for each user

ords_prods_merge['max_order']=ords_prods_merge.groupby(['user_id'])['order_number'].transform(ny.max)

In [17]:
ords_prods_merge.head(16)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   

    days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   
5                    19.0         196                  1          1   both   
6                    20.0         196                  1          1   both   
7                    14.0         196                  2          1   both   
8                     0.0         196                  4          1   both   
9                    30.0         196                  1          1   both   
10                    7.0         196                  2          0   both   
11                    8.0         196                  6          1   both   
12                    7.0         196                  1          1   both   
13                    7.0         196                  1          1   both   
14                   14.0         196                  1          1   both   
15                    6.0         196                  3          0   both   

   product_name  aisle_id  department_id  prices        price_range  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda        77              7     9.0  Mid-range product   
12         Soda        77              7     9.0  Mid-range product   
13         Soda        77              7     9.0  Mid-range product   
14         Soda        77              7     9.0  Mid-range product   
15         Soda        77              7     9.0  Mid-range product   

       busiest_day     busiest_days busiest_period_of_day  max_order  
0   Regularly busy   Regularly busy        Average orders         10  


In [18]:
ords_prods_merge.shape

(32404859, 19)

# Deriving Columns with loc()

In [19]:
# To create your flag, you’ll need some criteria. You can use the following:

# If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
# If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
# If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [20]:
# Based on the criteria listed in above section

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [21]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [22]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [23]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [24]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

In [25]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                    NaN         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

---

# Task Questions

---

# Step 1: Create New Notebook

#### Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [1]:
# Done

# Step 2: Aggregated Mean of Entire Dataframe

#### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [27]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3: Analyze Result and Compare with Subset Results

#### Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [28]:
# The subset

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

##### The mean of the subset's departments are actually higher than of the entire dataframe, save for department_id 17. The entire dataframe's means ranged from 15.46 to 22.90 while the subset ranged from 11.29 to 19.46. 

# Step 4: Create Loyalty Flag for Existing Customers

#### Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [29]:
# Done

# Step 5: Difference of Spending Habits in Loyalty Flag

#### The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [30]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

#### It's interesting that the highest average of spending is actually from the new customers, and the lowest amount is from the loyal customers. I don't know if it's because for the loyal customers, there are fewer actual loyal customers while the newer customers, there are a higher amount?

#### The maximum amount of 99999 seems a bit odd? Is it because of a bulk product?

In [31]:
# Checking to see if the 99999 exists or not.

df_max_price=ords_prods_merge[ords_prods_merge['prices']==99999]

In [32]:
df_max_price['product_name'].value_counts()

product_name
2 % Reduced Fat  Milk    698
Name: count, dtype: int64

In [33]:
df_max_price.head()

order_id  user_id  order_number  orders_day_of_week  \
29165516    183964      873             3                   0   
29165517   1851256      873             4                   6   
29165518   1915696     1893             1                   5   
29165519   2763293     1893             2                   4   
29165520   2564805     1893             4                   1   

          order_hour_of_day  days_since_last_order  product_id  \
29165516                 10                    7.0       33664   
29165517                 12                   13.0       33664   
29165518                 17                    NaN       33664   
29165519                 16                   13.0       33664   
29165520                 17                   30.0       33664   

          add_to_cart_order  reordered _merge           product_name  \
29165516                 11          0   both  2 % Reduced Fat  Milk   
29165517                  8          1   both  2 % Reduced Fat  Milk   
29165518                 10          0   both  2 % Reduced Fat  Milk   
29165519                  6          1   both  2 % Reduced Fat  Milk   
29165520                  3          1   both  2 % Reduced Fat  Milk   

          aisle_id  department_id   prices         price_range  \
29165516        84             16  99999.0  High-range product   
29165517        84             16  99999.0  High-range product   
29165518        84             16  99999.0  High-range product   
29165519        84             16  99999.0  High-range product   
29165520        84             16  99999.0  High-range product   

             busiest_day     busiest_days busiest_period_of_day  max_order  \
29165516     Busiest day     Busiest days           Most orders          8   
29165517  Regularly busy   Regularly busy           Most orders          8   
29165518  Regularly busy   Regularly busy        Average orders          6   
29165519      Least busy  Least busy days           Most orders          6   
29165520  Regularly busy     Busiest days        Average orders          6   

          loyalty_flag  
29165516  New customer  
29165517  New customer  
29165518  New customer  
29165519  New customer  
29165520  New customer

#### So it looks like the most expensive product is 2% milk, which is bought by all the customers

In [34]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                    NaN         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

#### Dealing with the Outliers

In [35]:
#Checking for threshold of normal data 

ords_prods_merge.loc[ords_prods_merge['prices'] > 100]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_last_order  product_id  \
10030345                 14                    5.0       21553   
10030346                 16                    4.0       21553   
10030347                 21                   13.0       21553   
10030348                  8                   12.0       21553   
10030349                 19                   30.0       21553   
...                     ...                    ...         ...   
29166209                  8                    4.0       33664   
29166210                  9                    2.0       33664   
29166211                 15                    8.0       33664   
29166212                 20                    4.0       33664   
29166213                 11                    4.0       33664   

          add_to_cart_order  reordered _merge  \
10030345                  5          0   both   
10030346                  3          1   both   
10030347                  6          0   both   
10030348                  3          1   both   
10030349                  9          0   both   
...                     ...        ...    ...   
29166209                  1          0   both   
29166210                  1          1   both   
29166211                 25          0   both   
29166212                  8          0   both   
29166213                 23          1   both   

                              product_name  aisle_id  department_id   prices  \
10030345  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030346  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030347  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030348  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030349  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
...                                    ...       ...            ...      ...   
29166209             2 % Reduced Fat  Milk        84             16  99999.0   
29166210             2 % Reduced Fat  Milk        84             16  99999.0   
29166211             2 % Reduced Fat  Milk        84             16  99999.0   
29166212             2 % Reduced Fat  Milk        84             16  99999.0   
29166213             2 % Reduced Fat  Milk        84             16  99999.0   

                 price_range     busiest_day     busiest_days  \
10030345  High-range product  Regularly busy   Regularly busy   
10030346  High-range product  Regularly busy   Regularly busy   
10030347  High-range product  Regularly busy   Regularly busy   
10030348  High-range product     Busiest day     Busiest days   
10030349  High-range product  Regularly busy     Busiest days   
...                      ...             ...              ...   
29166209  High-range product     Busiest day     Busiest days   
29166210  High-range product     Busiest day     Busiest days   
29166211  High-range product  Regularly busy  Least busy days   
29166212  High-range product  Regularly busy  Least busy days   
29166213  High-range product     Busiest day     Busiest days   

         busiest_period_of_day  max_order      loyalty_flag  
10030345           Most orders         40  

In [36]:
# Turning outliers into NaNs

ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = ny.nan

In [37]:
# Running max check

ords_prods_merge['prices'].max()

25.0

# Step 6: Different Types of Spenders

#### The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
#### a) If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
#### b) If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [39]:
# Creating “average_price” column using transform()

ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(ny.mean)

In [40]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                    NaN         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

   ... aisle_id  department_id  prices        price_range     busiest_day  \
0  ...       77              7     9.0  Mid-range product  Regularly busy   
1  ...       77              7     9.0  Mid-range product  Regularly busy   
2  ...       77              7     9.0  Mid-range product  Regularly busy   
3  ...       77              7     9.0  Mid-range product      Least busy   
4  ...       77              7     9.0  Mid-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag average_price  
0   Regularly busy        Average orders        10  New customer      6.367797  
1  Least busy days        Average orders        10  New customer      6.367797  
2  Least busy days           Most orders        10  New customer      6.367797  
3  Least busy days        Average orders        10  New customer      6.367797  
4  Least busy days           Most orders        10  New customer      6.367797  

[5 rows x 21 columns]

In [41]:
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [42]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [43]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                    NaN         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

   ... department_id  prices        price_range     busiest_day  \
0  ...             7     9.0  Mid-range product  Regularly busy   
1  ...             7     9.0  Mid-range product  Regularly busy   
2  ...             7     9.0  Mid-range product  Regularly busy   
3  ...             7     9.0  Mid-range product      Least busy   
4  ...             7     9.0  Mid-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1  Least busy days        Average orders        10  New customer   
2  Least busy days           Most orders        10  New customer   
3  Least busy days        Average orders        10  New customer   
4  Least busy days           Most orders        10  New customer   

   average_price spending_flag  
0       6.367797   Low spender  
1       6.367797   Low spender  
2       6.367797   Low spender  
3       6.367797   Low spender  
4       6.367797   Low spender  

[5 rows x 22 columns]

In [44]:
# Check frequency 

ords_prods_merge['spending_flag'].value_counts(dropna=False)

spending_flag
Low spender     32285131
High spender      119728
Name: count, dtype: int64

#### There are more low spenders than high spenders.

# Step 7: Frequent vs. Non-frequent Customers

#### In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
#### a.) If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
#### b.) If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
#### c.) If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [46]:
#Renaming my days since last order column to match the task instructions 

ords_prods_merge.rename(columns = {'days_since_last_order' : 'days_since_prior_order'}, inplace = True)

In [47]:
# Creating median_prior_orders column using transform()

ords_prods_merge['median_prior_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(ny.median)

In [48]:
#Checking if the renaming worked

ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... prices        price_range     busiest_day     busiest_days  \
0  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
1  ...    9.0  Mid-range product  Regularly busy  Least busy days   
2  ...    9.0  Mid-range product  Regularly busy  Least busy days   
3  ...    9.0  Mid-range product      Least busy  Least busy days   
4  ...    9.0  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag average_price  spending_flag  \
0        Average orders        10  New customer      6.367797    Low spender   
1        Average orders        10  New customer      6.367797    Low spender   
2           Most orders        10  New customer      6.367797    Low spender   
3        Average orders        10  New customer      6.367797    Low spender   
4           Most orders        10  New customer      6.367797    Low spender   

  median_prior_orders  
0                20.5  
1                20.5  
2                20.5  
3                20.5  
4                20.5  

[5 rows x 23 columns]

In [49]:
# Creating a frequency flag based on the criteria above 

ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Frequent customer'

In [50]:
ords_prods_merge.loc[(ords_prods_merge['median_prior_orders'] > 10) & (ords_prods_merge['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [51]:
ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Non-frequent customer'

In [52]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ...        price_range     busiest_day     busiest_days  \
0  ...  Mid-range product  Regularly busy   Regularly busy   
1  ...  Mid-range product  Regularly busy  Least busy days   
2  ...  Mid-range product  Regularly busy  Least busy days   
3  ...  Mid-range product      Least busy  Least busy days   
4  ...  Mid-range product      Least busy  Least busy days   

   busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0         Average orders        10  New customer      6.367797   Low spender   
1         Average orders        10  New customer      6.367797   Low spender   
2            Most orders        10  New customer      6.367797   Low spender   
3         Average orders        10  New customer      6.367797   Low spender   
4            Most orders        10  New customer      6.367797   Low spender   

   median_prior_orders order_frequency_flag  
0                 20.5    Frequent customer  
1                 20.5    Frequent customer  
2                 20.5    Frequent customer  
3                 20.5    Frequent customer  
4                 20.5    Frequent customer  

[5 rows x 24 columns]

In [53]:
# Check frequency 

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Non-frequent customer    21559853
Regular customer          7208564
Frequent customer         3636437
NaN                             5
Name: count, dtype: int64

There are 5 NaN values, I want to check what these are for.

In [54]:
# Check the null values

ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'median_prior_orders', 'order_frequency_flag']]

user_id  days_since_prior_order  median_prior_orders  \
13645692   159838                     NaN                  NaN   
17251990   159838                     NaN                  NaN   
17622767   159838                     NaN                  NaN   
24138593   159838                     NaN                  NaN   
25880002   159838                     NaN                  NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN

It seems all the NaN values are coming from one user, which is user_id 159838

In [55]:
# Check user_id 159838

ords_prods_merge[ords_prods_merge['user_id'] == 159838]

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered _merge  ...        price_range  \
13645692                  3          0   both  ...  Mid-range product   
17251990                  6          0   both  ...  Mid-range product   
17622767                  2          0   both  ...  Low-range product   
24138593                  5          0   both  ...  Mid-range product   
25880002                  1          0   both  ...  Low-range product   

          busiest_day  busiest_days  busiest_period_of_day max_order  \
13645692  Busiest day  Busiest days         Average orders         1   
17251990  Busiest day  Busiest days         Average orders         1   
17622767  Busiest day  Busiest days         Average orders         1   
24138593  Busiest day  Busiest days         Average orders         1   
25880002  Busiest day  Busiest days         Average orders         1   

          loyalty_flag average_price spending_flag  median_prior_orders  \
13645692  New customer          7.42   Low spender                  NaN   
17251990  New customer          7.42   Low spender                  NaN   
17622767  New customer          7.42   Low spender                  NaN   
24138593  New customer          7.42   Low spender                  NaN   
25880002  New customer          7.42   Low spender                  NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN  

[5 rows x 24 columns]

#### It appears that the customer with user_id 159838 is a new customer, therefore it is best to leave the NaN as is since it's still useful data.

In [56]:
# Check shape

ords_prods_merge.shape

(32404859, 24)

# Exporting to Pickle File

In [57]:
ords_prods_merge.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))